File: NPE.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading;
using SFML;
using SFML.Graphics;
using SFML.Window;

namespace CastleEscape
{
    /// <summary>
    /// Represents a non-player entity.
    /// 
    /// Author: Dennis Honeyman
    /// </summary>
    class NPE
    {
        private Action<Player> interactFunc;
        private Texture2D texture;
        private int xPos, yPos;
        private Game game;

        public NPE(Game game)
        {
            this.game = game;
        }

        public int XPos
        {
            get { return xPos; }
        }

        public int YPos
        {
            get { return yPos; }
        }

        public void DrawBase(SpriteBatch spriteBatch, ScriptableMap map, int x, int y)
        {
            if (texture == null)
                return;
            var sourceRect = new Rectangle(0, texture.Height - map.TileSize, map.TileSize, map.TileSize);
            var destRect = new Rectangle(x + xPos * map.TileSize, y + yPos * map.TileSize, map.TileSize, map.TileSize);
            spriteBatch.Draw(texture, destRect, sourceRect, Color.White);
        }

        public void DrawTop(SpriteBatch spriteBatch, ScriptableMap map, int x, int y)
        {
            if (texture == null)
                return;

            int renderX = x + xPos * map.TileSize;
            int renderY = y + yPos * map.TileSize - (texture.Height - map.TileSize);
            var sourceRect = new Rectangle(0, 0, map.TileSize, texture.Height - map.TileSize);
            var destRect = new Rectangle(renderX, renderY, map.TileSize, texture.Height - map.TileSize);
            spriteBatch.Draw(texture, destRect, sourceRect, Color.White);
        }

        public void SetPosition(int x, int y)
        {
            xPos = x;
            yPos = y;
        }

        public void SetTexture(string textureName)
        {
            texture = game.Content.Load<Texture2D>(textureName);
        }

        public void SetInteractFunc(Action<Player> func)
        {
            interactFunc = func;
        }

        public void Interact(Player player)
        {
            var thread = new Thread(new ParameterizedThreadStart(interact));
            thread.Start(player);
        }

        private void interact(object player)
        {
            interactFunc((Player)player);
        }
    }
}

File: Source/NPE.cc
#include "NPE.hh"

#include <iostream>
#include <string>
using namespace std;

#include <SFML/Graphics.hpp>

namespace CastleEscape {

NPE::NPE() {
	//TODO
}

int NPE::GetXPos() {
	return xPos_;
}

int NPE::GetYPos() {
	return yPos_;
}

void NPE::SetPosition(int x, int y) {
	xPos_ = x;
	yPos_ = y;
}

void NPE::SetTexture(string textureName) {
	texture_.LoadFromFile("Content/Graphics/" + textureName);
	texture_.SetSmooth(false);
}

void NPE::Draw(sf::RenderWindow& window, int x, int y) {
	sf::Sprite sprite(texture_);
	sprite.SetPosition(x, y);
	window.Draw(sprite);
}

ostream& operator<<(ostream& os, const NPE& npe) {
	return os << "NPE: (" << npe.xPos_ << "," << npe.yPos_ << ")";
}

} // namespace CastleEscape
